### ISSUE TITLE
AAM Timeout not working

### ISSUE THREAD

**Issue by RobinFrcd** [epam/Indigo#318]
Hi,
I'm working on pretty large reactions and the `aam-timeout` option for the automapping doesn't work as expected. On some large reactions, calling the `automap` can even make the kernel crash (OOM Kill).

Here is an example:

```python
from indigo import Indigo

rxn_smiles = '[F:1][C:2]([F:3])([C:4]([F:5])([F:6])[CH2:7][CH2:8][Si:9](Br)([CH2:10][CH2:11][C:12]([F:13])([F:14])[C:15]([F:16])([F:17])[C:18]([F:19])([F:20])[C:21]([F:22])([F:23])[C:24]([F:25])([F:26])[C:27]([F:28])([F:29])[F:30])[CH2:31][CH2:32][C:33]([F:34])([F:35])[C:36]([F:37])([F:38])[C:39]([F:40])([F:41])[C:42]([F:43])([F:44])[C:45]([F:46])([F:47])[C:48]([F:49])([F:50])[F:51])[C:52]([F:53])([F:54])[C:55]([F:56])([F:57])[C:58]([F:59])([F:60])[C:61]([F:62])([F:63])[F:64].[CH2:65]=[CH:66][CH2:67][OH:68]>CCN(CC)CC.C1CCOC1>[CH2:65]=[CH:66][CH2:67][O:68][Si:9]([CH2:10][CH2:11][C:12]([F:14])([F:13])[C:15]([F:17])([F:16])[C:18]([F:20])([F:19])[C:21]([F:23])([F:22])[C:24]([F:25])([F:26])[C:27]([F:30])([F:29])[F:28])([CH2:8][CH2:7][C:4]([F:5])([F:6])[C:2]([F:1])([F:3])[C:52]([F:53])([F:54])[C:55]([F:56])([F:57])[C:58]([F:59])([F:60])[C:61]([F:63])([F:62])[F:64])[CH2:31][CH2:32][C:33]([F:34])([F:35])[C:36]([F:37])([F:38])[C:39]([F:40])([F:41])[C:42]([F:43])([F:44])[C:45]([F:46])([F:47])[C:48]([F:50])([F:51])[F:49]'

mapper = Indigo()
mapper.setOption("timeout", 500)
mapper.setOption("aam-timeout", 500)
mapped_reaction = mapper.loadReaction(rxn_smiles)

mapped_reaction.automap(
    "discard ignore_charges ignore_radicals ignore_isotopes ignore_valence"
)
mapped_smiles = mapped_reaction.smiles().split(' ')[0]
```

<img width="1271" alt="image" src="https://user-images.githubusercontent.com/29704178/106634299-f32d0d00-657f-11eb-8035-f6f77a7937b4.png">

I understand why this reaction can be very hard to map, but I don't understand why the timeout doesn't work here.

Is there an option/function somewhere to predict whether the automap will be able to work normally on a reaction ? I'd rather skip the mapping of this kind of reaction than have an OOM error.

Thanks

**Comment by RobinFrcd** [comment]
This issue is really problematic for me, it often makes my app crash and there's no way to prevent this issue by catching an error because it just fills de RAM. 

Any idea how to identify a reaction that Indigo can't handle and just skip it ? Thanks!

**Comment by RobinFrcd** [comment]
It's not an API issue, I also get stuck in plain C++:

```c++
int main(void)
{
    const char* rxn_smiles =
        "[F:1][C:2]([F:3])([C:4]([F:5])([F:6])[CH2:7][CH2:8][Si:9](Br)([CH2:10][CH2:11][C:12]([F:13])([F:14])[C:15]([F:16])([F:17])[C:18]([F:19])([F:20])[C:21]([F:22])([F:23])[C:24]([F:25])([F:26])[C:27]([F:28])([F:29])[F:30])[CH2:31][CH2:32][C:33]([F:34])([F:35])[C:36]([F:37])([F:38])[C:39]([F:40])([F:41])[C:42]([F:43])([F:44])[C:45]([F:46])([F:47])[C:48]([F:49])([F:50])[F:51])[C:52]([F:53])([F:54])[C:55]([F:56])([F:57])[C:58]([F:59])([F:60])[C:61]([F:62])([F:63])[F:64].[CH2:65]=[CH:66][CH2:67][OH:68]>CCN(CC)CC.C1CCOC1>[CH2:65]=[CH:66][CH2:67][O:68][Si:9]([CH2:10][CH2:11][C:12]([F:14])([F:13])[C:15]([F:17])([F:16])[C:18]([F:20])([F:19])[C:21]([F:23])([F:22])[C:24]([F:25])([F:26])[C:27]([F:30])([F:29])[F:28])([CH2:8][CH2:7][C:4]([F:5])([F:6])[C:2]([F:1])([F:3])[C:52]([F:53])([F:54])[C:55]([F:56])([F:57])[C:58]([F:59])([F:60])[C:61]([F:63])([F:62])[F:64])[CH2:31][CH2:32][C:33]([F:34])([F:35])[C:36]([F:37])([F:38])[C:39]([F:40])([F:41])[C:42]([F:43])([F:44])[C:45]([F:46])([F:47])[C:48]([F:50])([F:51])[F:49]";
    indigoSetOptionFloat("aam-timeout", 10000);
    int rxn = indigoLoadReactionFromString(rxn_smiles);
    indigoAutomap(rxn, "discard ignore_charges ignore_radicals ignore_isotopes ignore_valence");
}  
```

Looks like it gets stuck in `_createReactionMap()`

**Comment by RobinFrcd** [comment]
@mkviatkovskii Hello there! I just submitted a [PR](https://github.com/epam/Indigo/pull/395) which fixes the issue for the provided example. :)

**Comment by RobinFrcd** [comment]
Any news here, please ? :)

**Comment by AlexanderSavelyev** [comment]
It seems there are some tests failures and also conflicts in the PR. I will check

**Comment by AlexanderSavelyev** [comment]
resolved the bug via #414 
I have changed the behavior to raise an exception on timeout (before the method was trying to complete the atom mapping even after the timeout). Right now it should be more time-precise when the aam stops. 
As for OOM - need to create additional spike
